src/passes/MultiMemoryLowering.cpp

```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Condensing Multi-Memories into a single memory for browsers that don’t
+// support multiple memories.
+//
+
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "wasm-builder.h"
+#include <pass.h>
+#include <wasm.h>
+
+namespace wasm {
+
+struct MultiMemoryLowering : public Pass {
+  Module* wasm = nullptr;
+  // The name of the single memory that exists after this pass is run
+  Name combinedMemory;
+  // The type of the single memory
+  Type pointerType;
+  // Used to indicate the type of the single memory when creating instructions
+  // (memory.grow, memory.size) for that memory
+  Builder::MemoryInfo memoryInfo;
+  // If the combined memory is shared
+  bool isShared;
+  // The initial page size of the combined memory
+  Address totalInitialPages;
+  // The max page size of the combined memory
+  Address totalMaxPages;
+  // There is no offset for the first memory, so offsetGlobalNames will always
+  // have a size that is one less than the count of memories at the time this
+  // pass is run. Use helper getOffsetGlobal(Index) to index the vector
+  // conveniently without having to manipulate the index directly
+  std::vector<Name> offsetGlobalNames;
+  // Maps from the name of the memory to its index as seen in the
+  // module->memories vector
+  std::unordered_map<Name, Index> memoryIdxMap;
+  // A vector of the memory size function names that were created proactively
+  // for each memory
+  std::vector<Name> memorySizeNames;
+  // A vector of the memory grow functions that were created proactively for
+  // each memory
+  std::vector<Name> memoryGrowNames;
+
+  void run(Module* module) override {
+    // Disabling multi-memories so that the target features esction in the
+    // emitted module does not report the use of MultiMemories. Also prevents
+    // later passes from adding additional memories.
```

Comment:
Maybe move this comment to the top comment at the start of the file? It's important for people using the pass to know it I think.

Reply:
Good idea, thanks!